agent/consul/gateways/controller_gateways.go

```diff
@@ -2,27 +2,166 @@ package gateways
 	"github.com/hashicorp/consul/agent/consul/state"
 	"github.com/hashicorp/consul/agent/consul/stream"
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/go-hclog"
 )
 
 type apiGatewayReconciler struct {
-	fsm    *fsm.FSM
 	logger hclog.Logger
+	store  datastore.DataStore
+}
+
+func (r apiGatewayReconciler) retrieveAllRoutesFromStore() ([]structs.BoundRoute, error) {
+	tcpRoutes, err := r.store.GetConfigEntriesByKind(structs.TCPRoute)
+	if err != nil {
+		return nil, err
+	}
+
+	//TODO not implemented
+	//httpRoutes, err := r.store.GetConfigEntriesByKind(structs.HTTPRoute)
+	//if err != nil {
+	//	return nil, err
+	//}
+
+	routes := []structs.BoundRoute{}
+	for _, r := range tcpRoutes {
+		if r == nil {
+			continue
+		}
+		routes = append(routes, r.(*structs.TCPRouteConfigEntry))
+	}
+	//TODO not implemented
+	//for _, r := range httpRoutes {
+	//	routes = append(routes, r.(*structs.HTTPRouteConfigEntry))
+	//}
+	fmt.Println(routes)
+	return routes, nil
 }
 
 func (r apiGatewayReconciler) Reconcile(ctx context.Context, req controller.Request) error {
+	//TODO thomases wrapper function
+	entry, err := r.store.GetConfigEntry(req.Kind, req.Name, req.Meta)
+	if err != nil {
+		return err
+	}
+
+	if entry == nil {
+		r.logger.Error("cleaning up deleted gateway object", "request", req)
+		if err := r.store.Delete(&structs.BoundAPIGatewayConfigEntry{
+			Kind:           structs.BoundAPIGateway,
+			Name:           req.Name,
+			EnterpriseMeta: *req.Meta,
+		}); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	gatewayEntry := entry.(*structs.APIGatewayConfigEntry)
+	//TODO is this what needs to happen for the validation step
+	err = gatewayEntry.Validate()
+	if err != nil {
+		r.logger.Debug("persisting gateway status", "gateway", gatewayEntry)
+		if err := r.store.UpdateStatus(gatewayEntry); err != nil {
+			return err
+		}
+		return err
+	}
+
+	var boundGatewayEntry *structs.BoundAPIGatewayConfigEntry
+	boundEntry, err := r.store.GetConfigEntry(structs.BoundAPIGateway, req.Name, req.Meta)
+	if err != nil {
+		return err
+	}
+	if boundEntry == nil {
+		boundGatewayEntry = &structs.BoundAPIGatewayConfigEntry{
+			Kind:           structs.BoundAPIGateway,
+			Name:           gatewayEntry.Name,
+			EnterpriseMeta: gatewayEntry.EnterpriseMeta,
+		}
+	} else {
+		boundGatewayEntry = boundEntry.(*structs.BoundAPIGatewayConfigEntry)
+	}
+
+	r.logger.Debug("started reconciling gateway")
+
+	routes, err := r.retrieveAllRoutesFromStore()
+
+	if err != nil {
+		return err
+	}
+	boundGateways, routeErrors := BindRoutesToGateways(wrapGatewaysInSlice(boundGatewayEntry), routes...)
+
+	if len(boundGateways) > 1 {
+		r.logger.Warn("imlpementation error, state should be impossible in gateway reconciler")
+		return errors.New("incorrect number of gateways bound")
+	} else if len(boundGateways) == 0 && len(routeErrors) == 0 {
+		r.logger.Debug("gateway had no updates to make ")
+		return nil
+	}
```

Comment:
In the context of debugging, I think it would be helpful to give more metadata about the gateway when creating the errors and debug logging. 

That first state should be impossible here, but users do sometimes make the impossible possible.

I might try this instead:

```suggestion
	if len(boundGateways) > 1 {
	    err := fmt.Errorf("bind returned more gateways (%d) than it was given (1)", len(boundGateways))
	    r.logger.Errorf("API Gateway Reconciler failed to reconcile: %v", err)
		return err
	} 
	
	if len(boundGateways) == 0 && len(routeErrors) == 0 {
		r.logger.Debugf("API Gateway Reconciler: gateway %s reconciled without updates.")
		return nil
	}
```